File: src/main/scala/com/bitdecay/tictactoe/scala/action/WinAction.scala
package com.bitdecay.tictactoe.scala.action

import com.bitdecay.board.Action
import com.bitdecay.board.utils.GameBoardException
import com.bitdecay.tictactoe.scala.State

class WinAction extends Action[State] {
  private var side:Int = 0
  def setSide(side:Int) = if (side == 0) throw new GameBoardException("Side cannot be 0") else this.side = side

  override def innerApply(t: State): State = t.copy(winner=side, finished=true)

  override def innerUnapply(t: State): State = t.copy(winner=0, finished=false)

  override def description(): String = ???

  override def serialize(): String = ???

  override def deserialize(s: String) = ???
}

File: src/main/java/com/bitdecay/tictactoe/action/WinAction.java
package com.bitdecay.tictactoe.action;


import com.bitdecay.board.Action;
import com.bitdecay.board.utils.GameBoardException;
import com.bitdecay.tictactoe.State;

public class WinAction extends Action<State> {
    private int side = 0;

    public void setSide(int side){
        if (side == 0) throw new GameBoardException("Side cannot be 0");
        this.side = side;
    }

    @Override
    protected State innerApply(State state) {
        State next = (State) state.clone();
        next.winner = side;
        next.finished = true;
        return next;
    }

    @Override
    protected State innerUnapply(State state) {
        State next = (State) state.clone();
        next.winner = 0;
        next.finished = false;
        return next;
    }

    @Override
    public String description() {
        return null;
    }

    @Override
    public String serialize() {
        return null;
    }

    @Override
    public Object deserialize(String s) {
        return null;
    }
}
